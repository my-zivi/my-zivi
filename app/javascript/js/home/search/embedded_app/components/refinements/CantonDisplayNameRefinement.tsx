import { RefinementList } from 'react-instantsearch-dom';
import React from 'preact/compat';
import { flatten, orderBy, partition } from 'lodash';
import defaultTranslations from './defaultTranslations';

const transformItems = (items) => (
  flatten(
    partition(items, ['isRefined', true]).map((partitioned) => orderBy(partitioned, ['label', 'asc'])),
  )
);

const CantonDisplayNameRefinement: React.FunctionComponent = () => (
  <RefinementList
    attribute="canton_display_name"
    searchable={true}
    limit={5}
    showMore={true}
    transformItems={
      transformItems
    }
    translations={defaultTranslations}
  />
);

export default CantonDisplayNameRefinement;
