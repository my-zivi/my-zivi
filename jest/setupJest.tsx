/* eslint-disable @typescript-eslint/no-explicit-any */
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-preact-pure';
import React, { h, Fragment } from 'preact';
import '../app/javascript/js/shared/globals.d';

(global as any).h = h;
(global as any).Fragment = Fragment;

const globalConfig: MyZiviGlobal = {
  locale: 'de',
  translations: {
    servicesOverview: {
      title: 'My Services Overview',
      errors: {
        cannotLoad: 'Konnte die EinsÃ¤tze nicht laden',
      },
    },
    search: {
      loadMore: 'string',
      noResultsFound: 'string',
      poweredBy: 'string',
      refinements: {
        canton: 'string',
        category: 'string',
        collapseButton: 'string',
        controls: {
          noResults: 'string',
          placeholder: 'string',
          reset: 'string',
          resetTitle: 'string',
          showLess: 'string',
          showMore: 'string',
          submitTitle: 'string',
        },
        language: 'string',
        languages: {
          french: 'string',
          german: 'string',
          italian: 'string',
        },
        month: {
          one: 'string',
          other: 'string',
        },
        minimumServiceMonths: 'string',
        priorityProgram: 'string',
        priorityProgramLabel: 'string',
      },
      ribbons: {
        featured: 'Feat.',
        featuredFull: 'Featured',
        new: 'Neu',
        urgent: 'Wichtig',
      },
      search: 'string',
      searchPlaceholder: 'string',
      statistics: 'string',
      searchView: {
        map: 'Karte',
        tiles: 'Raster',
      },
    } as any,
    loading: 'Laden...',
    error: 'Fehler',
  },
  paths: {
    servicesOverview: '/my/mock/services.json',
  },
  algolia: {
    apiKey: 'my-api-key',
    applicationId: 'my-application-id',
  },
};
(global as any).MyZivi = globalConfig;

jest.mock('@rails/ujs', () => jest.requireActual('js/shared/__mocks__/Rails'));

jest.mock('@fortawesome/react-fontawesome', () => {
  const component: React.FunctionComponent<any> = () => <></>;
  component.displayName = 'FontAwesomeIcon';
  return {
    FontAwesomeIcon: component,
  };
});

configure({ adapter: new Adapter() });
