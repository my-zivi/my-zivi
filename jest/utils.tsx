/* eslint-disable @typescript-eslint/no-explicit-any */
import { mount } from 'enzyme';
import { IApi } from 'js/shared/Api';
import preact from 'preact';
import { ApiContext } from 'js/shared/ApiProvider';
import React from 'preact/compat';
import { InstantSearch } from 'react-instantsearch-dom';
import { defer } from 'lodash';
import { mockClient } from '../app/javascript/js/tests/algolia/mocks';

type MountComponentType = (componentInstance: preact.JSX.Element, targetComponentName?: string) => any;
export const mountConsumer: MountComponentType = (componentInstance, targetComponentName) => {
  const root = mount(componentInstance);

  if (!targetComponentName) {
    return root;
  }

  return root.find(targetComponentName);
};

type FakeWrapInApiProviderType = (api: IApi) => (wrapped: preact.JSX.Element, component: string) => any;
export const fakeWrapInApiProvider: FakeWrapInApiProviderType = (api: IApi) => (
  (wrapped, component) => mountConsumer(
    (
      <ApiContext.Provider value={api}>
        {wrapped}
      </ApiContext.Provider>
    ),
    component,
  )
);

export const mountInstantSearchClient: MountComponentType = (componentInstance, targetComponentName?) => mountConsumer(
  <InstantSearch searchClient={mockClient} indexName="JobPosting">
    {componentInstance}
  </InstantSearch>,
  targetComponentName,
);

type DeferredCheck = (rendered: { update: () => void }, callback: () => void | Promise<void>) => Promise<void>;
export const deferredCheck: DeferredCheck = (rendered, callback) => new Promise<void>((resolve, reject) => {
  defer(() => {
    rendered.update();
    const result = callback();
    if (!result) {
      return resolve();
    }

    return result.then(resolve).catch(reject);
  });
});
