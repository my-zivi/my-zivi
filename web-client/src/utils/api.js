const BASE_URL = 'http://localhost:8000/api/';

// Is user logged in?
const isLoggedIn = () => {
  if (localStorage.getItem('jwtToken') !== null) {
    return true;
  } else {
    return false;
  }
};

// Is user an administrator?
export const isAdmin = () => {
  if (isLoggedIn()) {
    const jwtDecode = require('jwt-decode');
    const decodedToken = jwtDecode(localStorage.getItem('jwtToken'));
    return decodedToken.isAdmin;
  }
  return false;
};

// Verify that the fetched response is JSON
export function _verifyResponse(res) {
  let contentType = res.headers.get('content-type');

  if (contentType && contentType.indexOf('application/json') !== -1) {
    return res.json();
  } else {
    _handleError({ message: 'Response was not JSON' });
  }
}

const ApiService = { BASE_URL, isLoggedIn, isAdmin };
export default ApiService;
