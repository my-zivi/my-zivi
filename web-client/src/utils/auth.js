import moment from 'moment';
import jwtDecode from 'jwt-decode';
import Raven from 'raven-js';

// Is user logged in?
const isLoggedIn = () => {
  if (localStorage.getItem('jwtToken') !== null) {
    const decodedToken = jwtDecode(localStorage.getItem('jwtToken'));
    return moment.unix(decodedToken.exp).isAfter(moment());
  } else {
    return false;
  }
};

// Is user an administrator?
const isAdmin = () => {
  if (isLoggedIn()) {
    const decodedToken = jwtDecode(localStorage.getItem('jwtToken'));
    return decodedToken.isAdmin;
  }
  return false;
};

// Get logged in user id
const getUserId = () => {
  if (isLoggedIn()) {
    const decodedToken = jwtDecode(localStorage.getItem('jwtToken'));
    return decodedToken.sub;
  }
};

const setToken = token => {
  localStorage.setItem('jwtToken', token);
  Raven.setUserContext({ id: getUserId() });
};

// Verify that the fetched response is JSON
function _verifyResponse(res) {
  let contentType = res.headers.get('content-type');

  if (contentType && contentType.indexOf('application/json') !== -1) {
    return res.json();
  } else {
    _handleError({ message: 'Response was not JSON' });
  }
}

const Auth = { isLoggedIn, isAdmin, getUserId, setToken };
export default Auth;
