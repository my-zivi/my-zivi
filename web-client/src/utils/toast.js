import iziToast from 'izitoast/dist/js/iziToast.js';

function showSuccess(title, msg) {
  iziToast.show({
    id: 'save-success',
    theme: 'light',
    color: 'green',
    title: title,
    message: msg,
    position: 'topCenter',
    transitionIn: 'flipInX',
    transitionOut: 'flipOutX',
    progressBarColor: 'rgb(0, 255, 0)',
    layout: 2,
    timeout: 2000,
    onClose: function() {
      // console.info('onClose');
    },
  });
}

function showError(title, msg) {
  iziToast.show({
    id: 'save-success',
    theme: 'dark',
    color: 'red',
    title: title,
    message: msg,
    position: 'topCenter',
    transitionIn: 'flipInX',
    transitionOut: 'flipOutX',
    progressBarColor: 'rgb(255, 0, 0)',
    layout: 2,
    timeout: 4000,
    onClose: function() {
      // console.info('onClose');
    },
  });
}

// Export
const Toast = { showSuccess, showError };
export default Toast;
