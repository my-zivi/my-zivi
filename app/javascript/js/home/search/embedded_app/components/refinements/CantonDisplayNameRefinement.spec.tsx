import React from 'preact/compat';
import { deferredCheck, mountInstantSearchClient } from '../../../../../../../../jest/utils';
import CantonDisplayNameRefinement from './CantonDisplayNameRefinement';

describe('CantonDisplayNameRefinement', () => {
  const rendered = mountInstantSearchClient(<CantonDisplayNameRefinement />, 'CantonDisplayNameRefinement');

  it('renders correctly', () => deferredCheck(rendered, () => {
    expect(rendered).toMatchSnapshot();
  }));
});
