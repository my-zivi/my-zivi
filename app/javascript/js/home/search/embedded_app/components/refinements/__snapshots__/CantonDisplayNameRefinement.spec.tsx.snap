// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CantonDisplayNameRefinement renders correctly 1`] = `
<CantonDisplayNameRefinement>
  <ConnectorWrapper
    attribute="canton_display_name"
    limit={5}
    searchable={true}
    showMore={true}
    transformItems={[Function]}
    translations={
      Object {
        "noResults": "string",
        "placeholder": "string",
        "resetTitle": "string",
        "showMore": [Function],
        "submitTitle": "string",
      }
    }
  >
    <Consumer>
      <Consumer>
        <AlgoliaRefinementList(RefinementListWidget)
          attribute="canton_display_name"
          contextValue={
            Object {
              "createHrefForState": [Function],
              "mainTargetedIndex": "JobPosting",
              "onInternalStateUpdate": [Function],
              "onSearchForFacetValues": [Function],
              "onSearchParameters": [Function],
              "onSearchStateChange": [Function],
              "store": Object {
                "getState": [Function],
                "setState": [Function],
                "subscribe": [Function],
              },
              "widgetsManager": Object {
                "getWidgets": [Function],
                "registerWidget": [Function],
                "update": [Function],
              },
            }
          }
          limit={5}
          operator="or"
          searchable={true}
          showMore={true}
          showMoreLimit={20}
          transformItems={[Function]}
          translations={
            Object {
              "noResults": "string",
              "placeholder": "string",
              "resetTitle": "string",
              "showMore": [Function],
              "submitTitle": "string",
            }
          }
        >
          <RefinementListWidget
            attribute="canton_display_name"
            canRefine={false}
            createURL={[Function]}
            currentRefinement={Array []}
            isFromSearch={false}
            items={Array []}
            limit={5}
            operator="or"
            refine={[Function]}
            searchForItems={[Function]}
            searchable={true}
            showMore={true}
            showMoreLimit={20}
            transformItems={[Function]}
            translations={
              Object {
                "noResults": "string",
                "placeholder": "string",
                "resetTitle": "string",
                "showMore": [Function],
                "submitTitle": "string",
              }
            }
          >
            <PanelWrapper
              attribute="canton_display_name"
              canRefine={false}
              createURL={[Function]}
              currentRefinement={Array []}
              isFromSearch={false}
              items={Array []}
              limit={5}
              operator="or"
              refine={[Function]}
              searchForItems={[Function]}
              searchable={true}
              showMore={true}
              showMoreLimit={20}
              transformItems={[Function]}
              translations={
                Object {
                  "noResults": "string",
                  "placeholder": "string",
                  "resetTitle": "string",
                  "showMore": [Function],
                  "submitTitle": "string",
                }
              }
            >
              <Consumer>
                <PanelCallbackHandler
                  canRefine={false}
                  setCanRefine={[Function]}
                >
                  <Translatable(RefinementList)
                    attribute="canton_display_name"
                    canRefine={false}
                    createURL={[Function]}
                    currentRefinement={Array []}
                    isFromSearch={false}
                    items={Array []}
                    limit={5}
                    operator="or"
                    refine={[Function]}
                    searchForItems={[Function]}
                    searchable={true}
                    showMore={true}
                    showMoreLimit={20}
                    transformItems={[Function]}
                    translations={
                      Object {
                        "noResults": "string",
                        "placeholder": "string",
                        "resetTitle": "string",
                        "showMore": [Function],
                        "submitTitle": "string",
                      }
                    }
                  >
                    <RefinementList
                      attribute="canton_display_name"
                      canRefine={false}
                      class=""
                      className=""
                      createURL={[Function]}
                      currentRefinement={Array []}
                      isFromSearch={false}
                      items={Array []}
                      limit={5}
                      operator="or"
                      refine={[Function]}
                      searchForItems={[Function]}
                      searchable={true}
                      showMore={true}
                      showMoreLimit={20}
                      transformItems={[Function]}
                      translate={[Function]}
                      translations={
                        Object {
                          "noResults": "string",
                          "placeholder": "string",
                          "resetTitle": "string",
                          "showMore": [Function],
                          "submitTitle": "string",
                        }
                      }
                    >
                      <List
                        canRefine={false}
                        class=""
                        className=""
                        cx={[Function]}
                        isFromSearch={false}
                        items={Array []}
                        limit={5}
                        query=""
                        renderItem={[Function]}
                        searchForItems={[Function]}
                        searchable={true}
                        selectItem={[Function]}
                        showMore={true}
                        showMoreLimit={20}
                        translate={[Function]}
                      >
                        <div
                          className="ais-RefinementList ais-RefinementList--noRefinement"
                        >
                          <div
                            className="ais-RefinementList-searchBox"
                          >
                            <Translatable(SearchBox)
                              currentRefinement=""
                              focusShortcuts={Array []}
                              onSubmit={[Function]}
                              refine={[Function]}
                              translate={[Function]}
                            >
                              <SearchBox
                                autoFocus={false}
                                class=""
                                className=""
                                currentRefinement=""
                                focusShortcuts={Array []}
                                isSearchStalled={false}
                                loadingIndicator={
                                  <svg
                                    class="ais-SearchBox-loadingIcon"
                                    className="ais-SearchBox-loadingIcon"
                                    height="18"
                                    stroke="#444"
                                    viewBox="0 0 38 38"
                                    width="18"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <g
                                      fill="none"
                                      fill-rule="evenodd"
                                    >
                                      <g
                                        stroke-width="2"
                                        transform="translate(1 1)"
                                      >
                                        <circle
                                          cx="18"
                                          cy="18"
                                          r="18"
                                          stroke-opacity=".5"
                                        />
                                        <path
                                          d="M36 18c0-9.94-8.06-18-18-18"
                                        >
                                          <animateTransform
                                            attributeName="transform"
                                            dur="1s"
                                            from="0 18 18"
                                            repeatCount="indefinite"
                                            to="360 18 18"
                                            type="rotate"
                                          />
                                        </path>
                                      </g>
                                    </g>
                                  </svg>
                                }
                                onSubmit={[Function]}
                                refine={[Function]}
                                reset={
                                  <svg
                                    class="ais-SearchBox-resetIcon"
                                    className="ais-SearchBox-resetIcon"
                                    height="10"
                                    viewBox="0 0 20 20"
                                    width="10"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M8.114 10L.944 2.83 0 1.885 1.886 0l.943.943L10 8.113l7.17-7.17.944-.943L20 1.886l-.943.943-7.17 7.17 7.17 7.17.943.944L18.114 20l-.943-.943-7.17-7.17-7.17 7.17-.944.943L0 18.114l.943-.943L8.113 10z"
                                    />
                                  </svg>
                                }
                                searchAsYouType={true}
                                showLoadingIndicator={false}
                                submit={
                                  <svg
                                    class="ais-SearchBox-submitIcon"
                                    className="ais-SearchBox-submitIcon"
                                    height="10"
                                    viewBox="0 0 40 40"
                                    width="10"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M26.804 29.01c-2.832 2.34-6.465 3.746-10.426 3.746C7.333 32.756 0 25.424 0 16.378 0 7.333 7.333 0 16.378 0c9.046 0 16.378 7.333 16.378 16.378 0 3.96-1.406 7.594-3.746 10.426l10.534 10.534c.607.607.61 1.59-.004 2.202-.61.61-1.597.61-2.202.004L26.804 29.01zm-10.426.627c7.323 0 13.26-5.936 13.26-13.26 0-7.32-5.937-13.257-13.26-13.257C9.056 3.12 3.12 9.056 3.12 16.378c0 7.323 5.936 13.26 13.258 13.26z"
                                    />
                                  </svg>
                                }
                                translate={[Function]}
                              >
                                <div
                                  className="ais-SearchBox"
                                >
                                  <form
                                    action=""
                                    className="ais-SearchBox-form"
                                    noValidate={true}
                                    onReset={[Function]}
                                    onSubmit={[Function]}
                                    role="search"
                                  >
                                    <input
                                      autoCapitalize="off"
                                      autoComplete="off"
                                      autoCorrect="off"
                                      autoFocus={false}
                                      className="ais-SearchBox-input"
                                      maxLength="512"
                                      oninput={[Function]}
                                      placeholder="string"
                                      required={true}
                                      spellCheck="false"
                                      type="search"
                                      value=""
                                    />
                                    <button
                                      className="ais-SearchBox-submit"
                                      title="string"
                                      type="submit"
                                    >
                                      <svg
                                        className="ais-SearchBox-submitIcon"
                                        height="10"
                                        viewBox="0 0 40 40"
                                        width="10"
                                        xmlns="http://www.w3.org/2000/svg"
                                      >
                                        <path
                                          d="M26.804 29.01c-2.832 2.34-6.465 3.746-10.426 3.746C7.333 32.756 0 25.424 0 16.378 0 7.333 7.333 0 16.378 0c9.046 0 16.378 7.333 16.378 16.378 0 3.96-1.406 7.594-3.746 10.426l10.534 10.534c.607.607.61 1.59-.004 2.202-.61.61-1.597.61-2.202.004L26.804 29.01zm-10.426.627c7.323 0 13.26-5.936 13.26-13.26 0-7.32-5.937-13.257-13.26-13.257C9.056 3.12 3.12 9.056 3.12 16.378c0 7.323 5.936 13.26 13.258 13.26z"
                                        />
                                      </svg>
                                    </button>
                                    <button
                                      className="ais-SearchBox-reset"
                                      hidden={true}
                                      title="string"
                                      type="reset"
                                    >
                                      <svg
                                        className="ais-SearchBox-resetIcon"
                                        height="10"
                                        viewBox="0 0 20 20"
                                        width="10"
                                        xmlns="http://www.w3.org/2000/svg"
                                      >
                                        <path
                                          d="M8.114 10L.944 2.83 0 1.885 1.886 0l.943.943L10 8.113l7.17-7.17.944-.943L20 1.886l-.943.943-7.17 7.17 7.17 7.17.943.944L18.114 20l-.943-.943-7.17-7.17-7.17 7.17-.944.943L0 18.114l.943-.943L8.113 10z"
                                        />
                                      </svg>
                                    </button>
                                  </form>
                                </div>
                              </SearchBox>
                            </Translatable(SearchBox)>
                          </div>
                        </div>
                      </List>
                    </RefinementList>
                  </Translatable(RefinementList)>
                </PanelCallbackHandler>
              </Consumer>
            </PanelWrapper>
          </RefinementListWidget>
        </AlgoliaRefinementList(RefinementListWidget)>
      </Consumer>
    </Consumer>
  </ConnectorWrapper>
</CantonDisplayNameRefinement>
`;
